Translate 'Introducing Hooks' page (#28)
reactjs/pl.reactjs.org: content/docs/hooks-intro.md

---
id: hooks-intro
title: Wprowadzenie do Hooków
permalink: docs/hooks-intro.html
next: hooks-overview.html
---

*Hooki* są nowym dodatkiem w Reakcie 16.8. Pozwalają one używać stanu i innych funkcjonalności Reacta, bez użycia klas.

```js{4,5}
import React, { useState } from 'react';

function Example() {
  // Zadeklaruj nową zmienną stanu, którą nazwiemy "count"
  const [count, setCount] = useState(0);

  return (
    <div>

>       <p>Nacisnąłeś {count} razy</p>

Reviewer: Raczej powinniśmy pisać bezosobowo.

>Uwaga
>
>React 16.8.0 jest pierwszą wersją, która wspiera Hooki. Podczas aktualizacji nie zapomnij zaktualizować wszystkich paczek, w tym React DOM. React Native będzie wspierał Hooki w kolejnym, stabilnym wydaniu.

## Wprowadzenie wideo {#video-introduction}

Podczas konferencji "React Conf 2018" Sophie Alpert i Dan Abramov zaprezentowali po raz pierwszy Hooki. Następnie Ryan Florence zademonstrował, jak przepisać swoją aplikację, by móc ich używać. Wideo z konferencji zamieściliśmy poniżej:

<br>

<iframe width="650" height="366" src="//www.youtube.com/embed/dpw9EHDh2bM" frameborder="0" allowfullscreen></iframe>

## Bez krytycznych zmian {#no-breaking-changes}

Zanim przejdziemy dalej, zauważ że Hooki są:

* **Całkowicie opcjonalne.** Możesz wypróbować Hooki w kilku komponentach, bez przepisywania istniejącego kodu. Jeżeli jednak nie masz ochoty, nie musisz ich jeszcze stosować ani uczyć się o nich.
* **100% kompatybilne wstecznie.** Hooki nie zawierają żadnych zmian, które mogłyby zepsuć istniejący kod.
* **Dostępne już teraz.** Hooki są dostępne od wersji 16.8.0.

> **Nie ma planów na usunięcie klas z Reacta.**. Możesz przeczytać o strategii stopniowego wdrażania Hooków w [kolejnym podrozdziale](#gradual-adoption-strategy) tej strony.

Reviewer 2: ```suggestion
**Nie ma planów na usunięcie klas z Reacta.** Możesz przeczytać o strategii stopniowego wdrażania Hooków w [kolejnym podrozdziale](#gradual-adoption-strategy) tej strony.
```

Author: Kropka jest już w pogrubieniu

Reviewer 2: Chodzi mi ze u Ciebie sa 2 kropki .**.

Author: Rzeczywiście 🤦‍♂️ , zaraz poprawiam

**Jeżeli chcesz rozpocząć naukę o Hookach, [przejdź od razu do kolejnego rozdziału!](/docs/hooks-overview.html)** Możesz też kontynuować lekturę tego, aby dowiedzieć się, dlaczego w ogóle dodaliśmy Hooki, a także w jaki sposób będziemy je teraz stosować, bez potrzeby przepisywania naszych aplikacji.

## Motywacja {#motivation}

Hooki rozwiązują wiele, pozornie niepowiązanych ze sobą, problemów Reacta, na które natknęliśmy się podczas ponad pięciu lat pisania i utrzymywania dziesiątek tysięcy komponentów. Nie ważne, czy dopiero uczysz się Reacta, używasz go na co dzień, czy nawet preferujesz inną bibliotekę (o podobnym, komponentowym modelu działania) - możliwe, że natknąłeś się na część tych problemów.

### Współdzielenie logiki związanej ze stanem pomiędzy komponentami jest trudne {#its-hard-to-reuse-stateful-logic-between-components}

> React nie oferuje sposobu na "dołączenie" powtarzalnego zachowania do komponentu (na przykład, połączenie go z magazynem (ang. *store*)). Jeżeli pracujesz z Reactem już jakiś czas, najprawdopodobniej znasz wzorce, takie jak [właściwości renderujące (ang. *render props*)](/docs/render-props.html) i [komponenty wyższego rzędu (ang. *higher-order components*)](/docs/higher-order-components.html), które próbują rozwiązać ten problem. Wzorce te wymagają jednak modyfikacji komponentów w momencie ich użycia, co może być niewygodne i powodować, że kod jest trudniejszy w odbiorze. Jeśli spojrzysz na typową aplikację napisaną w Reakcie przy pomocy narzędzia React DevTools, najprawdopodobniej ujrzysz tam "piekło" komponentów opakowujących (ang. *wrapper component*), dostawców (ang. *providers*), konsumentów (ang. *consumers*), komponentów wyższego rzędu, właściwości renderujących i innych abstrakcji. Moglibyśmy, co prawda, [filtrować je w DevToolsach](https://github.com/facebook/react-devtools/pull/503), ale to tylko wskazuje na głębszy problem: React potrzebuje lepszego podstawowego mechanizmu do współdzielenia logiki związanej ze stanem.

Reviewer 2: ```suggestion
React nie oferuje sposobu na "dołączenie" powtarzalnego zachowania do komponentu (na przykład, połączenie go z magazynem (ang. *store*). Jeżeli pracujesz z Reactem już jakiś czas, najprawdopodobniej znasz wzorce, takie jak [właściwości renderujące (ang. *render props*)](/docs/render-props.html) i [komponenty wyższego rzędu (ang. *higher-order components*)](/docs/higher-order-components.html), które próbują rozwiązać ten problem. Wzorce te wymagają jednak modyfikacji komponentów w momencie ich użycia, co może być niewygodne i powodować, że kod jest trudniejszy w odbiorze. Jeśli spojrzysz na typową aplikację napisaną w Reakcie przy pomocy narzędzia React DevTools, najprawdopodobniej ujrzysz tam "piekło" komponentów opakowujących (ang. *wrapper component*), dostawców (ang. *providers*), konsumentów (ang. *consumers*), komponentów wyższego rzędu, właściwości renderujących i innych abstrakcji. Moglibyśmy, co prawda, [filtrować je w DevToolsach](https://github.com/facebook/react-devtools/pull/503), ale to tylko wskazuje na głębszy problem: React potrzebuje lepszego podstawowego mechanizmu do współdzielenia logiki związanej ze stanem.
```

Author: Tu są 2 nawiasy :). Możemy ten zewnętrzny zrobić kwadratowym, ale kierowałem się
https://sjp.pwn.pl/zasady/Uzycie-nawiasu-kwadratowego;629862.html
a konkretnie:
> c) Jeśli musimy użyć dwu nawiasów w odniesieniu do tego samego fragmentu, nawias kwadratowy może ujmować treści szersze, a nawias okrągły — węższe, np. (...) **Można jednak w takim wypadku stosować również podwójny nawias okrągły:**

Reviewer 2: Faktycznie! Nie zauważyłem tego pierwszego! INVALID

}
```

> Funkcja `useState` jest pierwszym "Hookiem", o którym będziemy się uczyć. Przykład ten jest jednak zaledwie zwiastunem. Nie przejmuj się, jeżeli nie ma to jeszcze większego sensu!

Reviewer: ```suggestion
Funkcja `useState` jest pierwszym "hookiem", o którym będziemy się uczyć. Przykład ten jest jednak zaledwie zwiastunem. Nie przejmuj się, jeżeli nie ma to jeszcze większego sensu!
```

Reviewer: Moim zdaniem lepiej zaczynać "hooki" małą literą.

Author: Chciałem być wierny oryginałowi w tej kwestii - z jakiegoś powodu wszędzie użyli wielkiej litery.

Reviewer: Może podeszli do tego jak do hasła reklamowego, nazwy własnej ;-) W języku polskim nadal, imho, lepiej pasuje pisać małą literą.

Author: Sprawdziłem Hiszpański, też jest wielką literą:
https://es.reactjs.org/docs/hooks-intro.html
Może skonsultować się z twórcami?

Reviewer: [Skonsultowałem się z "twórcami"](https://twitter.com/dan_abramov/status/1100933020536721408) :P i padła odpowiedź, że możemy pisać jak chcemy. Oni piszą wielką literą, bo chcieli nadać tej nazwie pewien stopień "specjalności".

Reviewer: Dla mnie dlatego pisanie od małej litery jest lepsze, bo to nie jest do końca ich wymyślona, własna nazwa. [Hooki istnieją w programowaniu od lat](https://pl.wikipedia.org/wiki/Hook_(informatyka)) i nie są niczym wyjątkowym - ot, kolejny mechanizm. Gdyby iść za ich schematem, to trzeba by pisać wielką literą "Props", "Lifecycle", "Components" - a tego tak nie piszemy :-)

Author: Ok, masz rację - zmienię na małą. Jak ktoś jest przeciwko, teraz jest dobry czas na sprzeciw ;).

Author: @cvqprs @iamarek @TeoTN @bartlomiejzuber @marcinostrouch @jakubdrozdek - może głosowanie? 🚀 - wielką, 👀 - małą

**[W kolejnym rozdziale](/docs/hooks-overview.html) możesz rozpocząć naukę o Hookach.** Tutaj wyjaśnimy, dlaczego dodaliśmy Hooki do Reacta i w jaki sposób pomogą ci one w pisaniu wspaniałych aplikacji.

> >Notka

Reviewer: Zwykle tłumaczyliśmy "Note" jako "Uwaga", co moim zdaniem lepiej nadaje się na nagłówek przypisu.

>
>React 16.8.0 jest pierwszą wersją, która wspiera Hooki. Podczas aktualizacji nie zapomnij zaktualizować wszystkich paczek, w tym React DOM. React Native będzie wspierał Hooki w kolejnym, stabilnym wydaniu.

## Wprowadzenie wideo {#video-introduction}

> Podczas React Conf 2018 Sophie Alpert i Dan Abramov zaprezentowali po raz pierwszy Hooki. Następnie Ryan Florence zademonstrował, jak przepisać swoją aplikację, by móc ich używać. Zobacz wideo tutaj:

Reviewer: ```suggestion
Podczas konferencji "React Conf 2018" Sophie Alpert i Dan Abramov zaprezentowali po raz pierwszy Hooki. Następnie Ryan Florence zademonstrował, jak przepisać swoją aplikację, by móc ich używać. Wideo z konferencji zamieściliśmy poniżej:
```

<br>

<iframe width="650" height="366" src="//www.youtube.com/embed/dpw9EHDh2bM" frameborder="0" allowfullscreen></iframe>

## Bez krytycznych zmian {#no-breaking-changes}

Zanim przejdziemy dalej, zauważ że Hooki są:

> * **Stosowane całkowicie wedle własnego uznania.** Możesz wypróbować Hooki w kilku komponentach, bez przepisywania istniejącego kodu. Jeżeli jednak nie masz ochoty, nie musisz ich jeszcze stosować, ani uczyć się o nich.

Reviewer: ```suggestion
* **Całkowicie opcjonalne.** Możesz wypróbować Hooki w kilku komponentach, bez przepisywania istniejącego kodu. Jeżeli jednak nie masz ochoty, nie musisz ich jeszcze stosować ani uczyć się o nich.
```
W tym kontekście, moim zdaniem, lepiej brzmi krótsza forma.

> * **100% wstecznie kompatybilne.** Hooki nie zawierają żadnych zmian, które mogłyby zepsuć istniejący kod.

Reviewer: ```suggestion
* **100% kompatybilne wstecznie.** Hooki nie zawierają żadnych zmian, które mogłyby zepsuć istniejący kod.
```

> * **Dostępne już teraz.** Hooki są dostępne wraz z wersją v16.8.0.

Reviewer: ```suggestion
* **Dostępne już teraz.** Hooki są dostępne od wersji 16.8.0.
```

**Nie ma planów na usunięcie klas z Reacta.**. Możesz przeczytać o strategii stopniowego wdrażania Hooków w [kolejnym podrozdziale](#gradual-adoption-strategy) tej strony.

> **Hooki nie zastępują twojej wiedzy na temat Reacta.** Zamiast tego wprowadzają bardziej bezpośredni interfejs API dla koncepcji Reacta, które już znasz: właściwości (ang. *props*), stanu, kontekstu, refów i cyklu życia (ang. *lifecycle*). Jak pokażemy dalej, Hooki pozwalają też na łączenie ich w nowy, niezwykle skuteczny sposób.

Reviewer: ```suggestion
**Hooki nie zastępują twojej wiedzy na temat Reacta.** Zamiast tego wprowadzają bardziej bezpośredni interfejs API dla mechanizmów Reacta, które już znasz: właściwości (ang. *props*), stanu, kontekstu, referencji (ang. *refs*) i cyklu życia (ang. *lifecycle*). Jak pokażemy dalej, Hooki pozwalają też na łączenie ich w nowy, niezwykle skuteczny sposób.
```

> **Jeżeli chcesz rozpocząć naukę o Hookach, [przejdź od razu do kolejnego rozdziału!](/docs/hooks-overview.html)** Możesz też kontynuować lekturę tego, aby dowiedzieć się, dlaczego w ogóle dodaliśmy Hooki. A także, w jaki sposób będziemy je teraz stosować, bez potrzeby przepisywania naszych aplikacji.

Reviewer: ```suggestion
**Jeżeli chcesz rozpocząć naukę o Hookach, [przejdź od razu do kolejnego rozdziału!](/docs/hooks-overview.html)** Możesz też kontynuować lekturę tego, aby dowiedzieć się, dlaczego w ogóle dodaliśmy Hooki, a także w jaki sposób będziemy je teraz stosować, bez potrzeby przepisywania naszych aplikacji.
```

## Motywacja {#motivation}

> Hooki rozwiązują wiele, pozornie niepowiązanych ze sobą problemów Reacta, na które natknęliśmy się podczas ponad pięciu lat pisania i utrzymywania dziesiątek tysięcy komponentów. Nie ważne, czy dopiero uczysz się Reacta, używasz go na co dzień, czy nawet preferujesz inną bibliotekę (o podobnym, komponentowym modelu działania), możliwe, że natknąłeś się na część tych problemów.

Reviewer: ```suggestion
Hooki rozwiązują wiele, pozornie niepowiązanych ze sobą, problemów Reacta, na które natknęliśmy się podczas ponad pięciu lat pisania i utrzymywania dziesiątek tysięcy komponentów. Nie ważne, czy dopiero uczysz się Reacta, używasz go na co dzień, czy nawet preferujesz inną bibliotekę (o podobnym, komponentowym modelu działania) - możliwe, że natknąłeś się na część tych problemów.
```

### Współdzielenie logiki związanej ze stanem pomiędzy komponentami jest trudne {#its-hard-to-reuse-stateful-logic-between-components}

> React nie oferuje sposobu na "dołączenie" powtarzalnego zachowania do komponentu (na przykład, połączenie go z magazynem (ang. *store*)). Jeżeli pracujesz z Reactem już jakiś czas, najprawdopodobniej znasz wzorce, takie jak [właściwość render (ang. *render props*)](/docs/render-props.html) i [komponenty wyższego rzędu (ang. *higher-order components*)](/docs/higher-order-components.html), które próbują rozwiązać ten problem. Wzorce te wymagają jednak  modyfikacji komponentów, w momencie ich użycia, co może być niewygodne i powodować, że kod jest trudniejszy w odbiorze. Jeśli spojrzysz na typową aplikację napisaną w Reakcie, w React DevToolsach, najprawdopodobniej ujrzysz tam "piekło" komponentów opakowujących (ang. *wrapper component*); dostawców (ang. *providers*), konsumentów (ang. *consumers*), komponentów wyższego rzędu, właściwości render i innych abstrakcji. Moglibyśmy, co prawda, [filtrować je w DevToolsach](https://github.com/facebook/react-devtools/pull/503), ale to tylko wskazuje na głębszy problem: React potrzebuje lepszego podstawowego mechanizmu do współdzielenia logiki związanej ze stanem.

Reviewer: ```suggestion
React nie oferuje sposobu na "dołączenie" powtarzalnego zachowania do komponentu (na przykład, połączenie go z magazynem (ang. *store*)). Jeżeli pracujesz z Reactem już jakiś czas, najprawdopodobniej znasz wzorce, takie jak [właściwości renderujące (ang. *render props*)](/docs/render-props.html) i [komponenty wyższego rzędu (ang. *higher-order components*)](/docs/higher-order-components.html), które próbują rozwiązać ten problem. Wzorce te wymagają jednak modyfikacji komponentów w momencie ich użycia, co może być niewygodne i powodować, że kod jest trudniejszy w odbiorze. Jeśli spojrzysz na typową aplikację napisaną w Reakcie przy pomocy narzędzia React DevTools, najprawdopodobniej ujrzysz tam "piekło" komponentów opakowujących (ang. *wrapper component*), dostawców (ang. *providers*), konsumentów (ang. *consumers*), komponentów wyższego rzędu, właściwości renderujących i innych abstrakcji. Moglibyśmy, co prawda, [filtrować je w DevToolsach](https://github.com/facebook/react-devtools/pull/503), ale to tylko wskazuje na głębszy problem: React potrzebuje lepszego podstawowego mechanizmu do współdzielenia logiki związanej ze stanem.
```
Props jako właściwości lub atrybuty - nadal do rozstrzygnięcia.
Oprócz tego wiem, że "render props" brzmi jak "właściwości renderowania", ale to jest pattern, który pozwala na przekazanie sposobu wyrenderowania komponentu przez komponent zarządzający samą logiką. Używa się do tego `children` albo konkretnego propsa, np. `render` czy `component`. Dlatego ja bym nadał temu trochę inny wydźwięk i nazwał "atrybutami renderującymi" (lub "właściwościami renderującymi").

Reviewer: Props jako właściwości lub atrybuty - nadal do rozstrzygnięcia.
Oprócz tego wiem, że "render props" brzmi jak "właściwości renderowania", ale to jest pattern, który pozwala na przekazanie sposobu wyrenderowania komponentu przez komponent zarządzający samą logiką. Używa się do tego `children` albo konkretnego propsa, np. `render` czy `component`. Dlatego ja bym nadał temu trochę inny wydźwięk i nazwał "atrybutami renderującymi" (lub "właściwościami renderującymi").

> React nie oferuje sposobu na "dołączenie" powtarzalnego zachowania do komponentu (na przykład, połączenie go z magazynem (ang. *store*)). Jeżeli pracujesz z Reactem już jakiś czas, najprawdopodobniej znasz wzorce, takie jak [właściwość render (ang. *render props*)](/docs/render-props.html) i [komponenty wyższego rzędu (ang. *higher-order components*)](/docs/higher-order-components.html), które próbują rozwiązać ten problem. Wzorce te wymagają jednak  modyfikacji komponentów, w momencie ich użycia, co może być niewygodne i powodować, że kod jest trudniejszy w odbiorze. Jeśli spojrzysz na typową aplikację napisaną w Reakcie, w React DevToolsach, najprawdopodobniej ujrzysz tam "piekło" komponentów opakowujących (ang. *wrapper component*); dostawców (ang. *providers*), konsumentów (ang. *consumers*), komponentów wyższego rzędu, właściwości render i innych abstrakcji. Moglibyśmy, co prawda, [filtrować je w DevToolsach](https://github.com/facebook/react-devtools/pull/503), ale to tylko wskazuje na głębszy problem: React potrzebuje lepszego podstawowego mechanizmu do współdzielenia logiki związanej ze stanem.

Reviewer: ```suggestion
React nie oferuje sposobu na "dołączenie" powtarzalnego zachowania do komponentu (na przykład, połączenie go z magazynem (ang. *store*)). Jeżeli pracujesz z Reactem już jakiś czas, najprawdopodobniej znasz wzorce, takie jak [właściwości renderujące (ang. *render props*)](/docs/render-props.html) i [komponenty wyższego rzędu (ang. *higher-order components*)](/docs/higher-order-components.html), które próbują rozwiązać ten problem. Wzorce te wymagają jednak modyfikacji komponentów w momencie ich użycia, co może być niewygodne i powodować, że kod jest trudniejszy w odbiorze. Jeśli spojrzysz na typową aplikację napisaną w Reakcie przy pomocy narzędzia React DevTools, najprawdopodobniej ujrzysz tam "piekło" komponentów opakowujących (ang. *wrapper component*), dostawców (ang. *providers*), konsumentów (ang. *consumers*), komponentów wyższego rzędu, właściwości renderujących i innych abstrakcji. Moglibyśmy, co prawda, [filtrować je w DevToolsach](https://github.com/facebook/react-devtools/pull/503), ale to tylko wskazuje na głębszy problem: React potrzebuje lepszego podstawowego mechanizmu do współdzielenia logiki związanej ze stanem.
```

Korzystając z Hooków, możesz wydzielić logikę związaną ze stanem z komponentu. Dzięki czemu, nie wymaga on zależności przy testowaniu i jest łatwy w ponownym wykorzystaniu. **Hooki pozwalają na ponowne użycie logiki związanej ze stanem, bez konieczności zmiany hierarchii komponentów.** Sprawia to, że dzielenie się Hookami pomiędzy wieloma komponentami lub ze społecznością jest proste.

> Omówimy ten temat szerzej w rozdziale [Tworzenie własnych Hooków](/docs/hooks-custom.html).

Reviewer: ```suggestion
Omówimy ten temat szerzej w rozdziale pt. ["Tworzenie własnych Hooków"](/docs/hooks-custom.html).
```

### Złożone komponenty stają się trudne do zrozumienia {#complex-components-become-hard-to-understand}

> Często musieliśmy utrzymywać komponenty, które z początku proste, urosły do rangi niemożliwego do utrzymania bałaganu logiki związanej ze stanem i efektów ubocznych (ang. *side effects*). Każda metoda cyklu życia zawiera zwykle mieszankę niepowiązanej ze sobą logiki. Na przykład, komponenty mogą pobierać dane w `componentDidMount` i `componentDidUpdate`. Jednakże metoda `componentDidMount` może też zawierać logikę, która tworzy słuchaczy zdarzeń. Następnie są oni czyszczeni w `componentWillUnmount`. Wzajemnie powiązany kod zostaje podzielony pomiędzy różne metody, a z kolei zupełnie niezwiązany ze sobą kod trafia do jednej. Otwiera to zbyt wielkie pole do błędów i niekonsekwencji.

Reviewer: ```suggestion
Często musieliśmy utrzymywać komponenty, które z początku proste, urosły do rangi niemożliwego do utrzymania bałaganu logiki związanej ze stanem i efektów ubocznych (ang. *side effects*). Każda metoda cyklu życia zawiera zwykle mieszankę niepowiązanej ze sobą logiki. Na przykład, komponenty mogą pobierać dane w `componentDidMount` i `componentDidUpdate`. Jednakże metoda `componentDidMount` może też zawierać logikę, która tworzy obserwatory zdarzeń (ang. *event listeners*). Następnie są one czyszczone w `componentWillUnmount`. Wzajemnie powiązany kod zostaje podzielony pomiędzy różne metody, a z kolei zupełnie niezwiązany ze sobą kod trafia do jednej. Otwiera to zbyt wielkie pole do błędów i niekonsekwencji.
```

Reviewer: Nie wiem co, ale coś mi nie pasuje w zdaniu "Otwiera to zbyt wielkie *pole* do błędów i niekonsekwencji". Czy to się tak mówi? :)

Author: Przez skojarzenie z _polem do nadużyć_, z którym się często spotykam

Reviewer: Albo z *polem do popisu* :-)

Reviewer: Może po prostu "sprzyja"?

Author: > Sprzyja to niekonsekwencji i popełnianiu błędów.
?

Reviewer: Pewnie :) 👍

Wielokrotnie zdarza się, że nie ma możliwości rozbicia tych komponentów na mniejsze części, ponieważ logika związana ze stanem jest już wszędzie. Trudno jest też je testować. Jest to jeden z powodów, dla których wielu woli połączyć Reacta z zewnętrzną biblioteką do zarządzania stanem. To jednak często wprowadza zbyt wiele abstrakcji, zmusza do skakania pomiędzy plikami i utrudnia ponowne wykorzystanie komponentów.

> Aby rozwiązać ten problem, **Hooki pozwalają podzielić komponent na mniejsze funkcje, bazując na powiązanych ze sobą częściach (takich jak tworzenie subskrypcji, czy pobieranie danych)**, zamiast wymuszać sztuczny podział, związany z metodami cyklu życia. Ewentualnie, aby uczynić to bardziej przewidywalnym, możesz też dzięki nim oddelegować zarządzanie lokalnym stanem komponentu do reduktora.

Reviewer: ```suggestion
Aby rozwiązać ten problem, **Hooki pozwalają podzielić komponent na mniejsze funkcje, bazując na powiązanych ze sobą częściach (takich jak tworzenie subskrypcji czy pobieranie danych)**, zamiast wymuszać sztuczny podział, związany z metodami cyklu życia. Ewentualnie, aby uczynić zachowanie komponentu bardziej przewidywalnym, możesz też dzięki hookom oddelegować zarządzanie lokalnym stanem komponentu do reduktora (ang. *reducer*).
```

> Szerzej omówimy to w rozdziale [Używanie Hooka Efektu](/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns).

Reviewer: Hook Efektu? Hmm, trochę mi to nie leży. Nie mam innego pomysłu na to, ale chętnie poczekam na zdanie innych osób.

Author: Muszę się z tym przespać ;)

Author: Jeżeli side effects przetłumaczymy na efekty uboczne, to ten hook pozwala na wykonywanie efektów ubocznych. Więc jest Hookiem Efektu :).

Author: Możemy też iść w stronę "skutku ubocznego"
https://pl.wikipedia.org/wiki/Skutek_uboczny_(informatyka)

Reviewer: Wiem, że w oryginale tego tam nie ma, ale czy możemy dodać ten drugi człon? "Hook Skutku Ubocznego"? Samo "Skutku" nic nie mówi...

Author: Samo _Effect_ też nie wiele mówi :). Nie chciałbym poprawiać twórców oryginału.

Author: Tłumacząc kolejną sekcję dochodzę do wniosku, że "efekt uboczny" będzie lepszy. Budzi on od razu skojarzenie z nazwą hooka - `useEffect`.

### Klasy dezorientują zarówno ludzi, jak i maszyny {#classes-confuse-both-people-and-machines}

> Oprócz tego, że jest przez nie trudniej ponownie wykorzystać i organizować kod, odkryliśmy, że klasy mogą stanowić dużą przeszkodę w nauce Reacta. Musisz rozumieć, jak działa `this` w JavaScripcie - a działa on tu zupełnie inaczej, niż w większości języków programowania. Musisz pamiętać o wiązaniu (ang. *bind*) funkcji obsługi zdarzeń (ang. *event handlers*). Bez wciąż niepotwierdzonych [propozycji składni](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/) kod jest bardzo rozwlekły. Ludzie generalnie nie mają problemu, ze zrozumieniem właściwości, stanu i kierunku przepływu danych z góry, do dołu, a jednak klasy wciąż stanowią pewne wyzwanie. Wybór pomiędzy funkcyjnymi, a klasowymi komponentami, jest często przyczyną sporów, nawet pomiędzy doświadczonymi programistami Reacta.

Reviewer: ```suggestion
Oprócz tego, że przez klasy jest trudniej ponownie wykorzystać i organizować kod, odkryliśmy, że mogą również stanowić dużą przeszkodę w nauce Reacta. Trzeba przecież rozumieć, jak działa `this` w JavaScripcie - a działa on tu zupełnie inaczej niż w większości języków programowania. Trzeba pamiętać o wiązaniu (ang. *bind*) funkcji obsługi zdarzeń (ang. *event handlers*). Bez, wciąż niepotwierdzonych, [propozycji składni](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/) kod jest bardzo rozwlekły. Ludzie generalnie nie mają problemu ze zrozumieniem właściwości, stanu i kierunku przepływu danych z góry do dołu, a jednak klasy wciąż stanowią pewne wyzwanie. Wybór pomiędzy funkcyjnymi a klasowymi komponentami jest często przyczyną sporów, nawet pomiędzy doświadczonymi programistami Reacta.
```

Reviewer: BTW. Nie wiem, dlaczego Git w niektórych sugestiach nie koloruje różnic... :(

Author: A co powiesz na taką formę:
```
Oprócz tego, że przez klasy trudniej jest ponownie wykorzystać i organizować kod, odkryliśmy, że mogą one również stanowić dużą przeszkodę w nauce Reacta. Trzeba przecież rozumieć, jak działa `this` w JavaScripcie - a działa on tu zupełnie inaczej niż w większości języków programowania. Trzeba pamiętać o wiązaniu (ang. *bind*) funkcji obsługi zdarzeń (ang. *event handlers*). Bez, wciąż niepotwierdzonych, [propozycji składni](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/) kod jest bardzo rozwlekły. Ludzie generalnie nie mają problemu ze zrozumieniem właściwości, stanu i kierunku przepływu danych z góry do dołu, a jednak klasy wciąż stanowią pewne wyzwanie. Wybór pomiędzy funkcyjnymi a klasowymi komponentami jest często przyczyną sporów, nawet pomiędzy doświadczonymi programistami Reacta.
```

![image](https://user-images.githubusercontent.com/7079786/53443987-48b02500-3a0d-11e9-9754-c39a598161d7.png)

> Ponadto, React jest dostępny od około pięciu lat i chcielibyśmy mieć pewność, że pozostanie on tak samo istotny przez kolejne pięć. Jak pokazał [Svelte](https://svelte.technology/), [Angular](https://angular.io/), [Glimmer](https://glimmerjs.com/) i inni [kompilacja komponentów z wyprzedzeniem (ang. *ahead-of-time*)](https://en.wikipedia.org/wiki/Ahead-of-time_compilation) ma ogromny potencjał. Szczególnie, jeśli nie jest ograniczona tylko do szablonów. Niedawno eksperymentowaliśmy z [wyliczaniem wartości komponentów](https://github.com/facebook/react/issues/7323) z użyciem [Prepacka](https://prepack.io/) i wstępnie zaobserwowaliśmy obiecujące rezultaty. Odkryliśmy jednak, że komponenty klasowe mogą zachęcić do nieumyślnego stosowania pewnych wzorców, które spowodują spowolnienie tych optymalizacji. Klasy już dzisiaj stanowią problem także dla naszych narzędzi programistycznych. Na przykład, nie za dobrze się minifikują, a gorące przeładowanie (ang. *hot reloading*) przez nie często zawodzi. Chcemy przedstawić interfejs API, który zwiększy prawdopodobieństwo tego, że kod będzie optymalizowany.

Reviewer: ```suggestion
Ponadto, React jest dostępny od około pięciu lat i chcielibyśmy mieć pewność, że pozostanie on tak samo istotny przez kolejne pięć. Jak pokazały biblioteki: [Svelte](https://svelte.technology/), [Angular](https://angular.io/), [Glimmer](https://glimmerjs.com/) i inne, [kompilacja komponentów z wyprzedzeniem (ang. *ahead-of-time*)](https://en.wikipedia.org/wiki/Ahead-of-time_compilation) ma ogromny potencjał. Szczególnie jeśli nie jest ograniczona tylko do szablonów. Niedawno eksperymentowaliśmy z [wyliczaniem wartości komponentów](https://github.com/facebook/react/issues/7323) z użyciem [Prepacka](https://prepack.io/) i wstępnie zaobserwowaliśmy obiecujące rezultaty. Odkryliśmy jednak, że komponenty klasowe mogą zachęcić do nieumyślnego stosowania pewnych wzorców, które spowodują spowolnienie tych optymalizacji. Klasy już dzisiaj stanowią problem także dla naszych narzędzi programistycznych. Na przykład, nie za dobrze się minifikują, a "gorące przeładowanie" (ang. *hot reloading*) przez nie często zawodzi. Chcemy przedstawić interfejs API, który zwiększy prawdopodobieństwo tego, że kod będzie optymalizowany.
```

Author: OK, ale przy okazji przeczytałem to zdanie
> Klasy już dzisiaj stanowią problem także dla naszych narzędzi programistycznych. 
Muszę je przepisać

Author: > Klasy już teraz stanowią problem dla naszych narzędzi programistycznych. Na przykład, nie za dobrze się minifikują i często zawodzi przez nie "gorące przeładowanie" (ang. *hot reloading*).

Reviewer: Ach to gorące przeładowanie :D

> Aby rozwiązać te problemy, **Hooki pozwalają na korzystanie z większej liczby funkcjonalności Reacta, bez użycia klas.** Koncepcyjnie, komponentom reactowym zawsze bliżej było do funkcji. Hooki zapewniają dostęp do funkcji, bez poświęcania praktycznej natury Reacta. Hooki zapewniają dostęp do imperatywnych furtek i nie wymagają nauki skomplikowanych technik programowania funkcjonalnego lub reaktywnego.

Reviewer: ```suggestion
Aby rozwiązać te problemy, **Hooki pozwalają na korzystanie z większej liczby funkcjonalności Reacta, bez użycia klas.** Koncepcyjnie, komponentom reactowym zawsze bliżej było do funkcji. Hooki zapewniają dostęp do funkcji bez poświęcania praktycznej natury Reacta. Hooki zapewniają dostęp do imperatywnych "furtek" i nie wymagają nauki skomplikowanych technik programowania funkcjonalnego lub reaktywnego.
```

>Przykłady
>

> >Rozdział [Hooki w pigułce](/docs/hooks-overview.html) jest dobrym miejscem, by zacząć naukę o Hookach

Reviewer: ```suggestion
>Rozdział pt. ["Hooki w pigułce"](/docs/hooks-overview.html) jest dobrym miejscem, by zacząć naukę o Hookach
```

## Strategia Stopniowego Wdrażania {#gradual-adoption-strategy}

>**TLDR: Nie ma planów na usunięcie klas z Reacta.**

Zdajemy sobie sprawę, że programiści Reacta są skupieni na dostarczaniu produktów i nie mają czasu przyglądać się każdemu nowemu interfejsowi API, który jest wypuszczany. Hooki są wielką nowością i być może lepiej będzie zaczekać na więcej przykładów i poradników, zanim rozważysz ich naukę lub wdrożenie.

> Rozumiemy też, że przy dodawaniu do Reacta nowego, podstawowego mechanizmu, poprzeczka jest postawiona niezwykle wysoko. Dla zainteresowanych, przygotowaliśmy [szczegółowe RFC](https://github.com/reactjs/rfcs/pull/68), które szczegółowo zgłębia nasze motywy i rzuca dodatkowe światło na konkretne decyzje projektowe i obecny stan techniki.

Reviewer: ```suggestion
Rozumiemy też, że przy dodawaniu do Reacta nowego, podstawowego mechanizmu poprzeczka została postawiona niezwykle wysoko. Dla zainteresowanych przygotowaliśmy [szczegółowy RFC](https://github.com/reactjs/rfcs/pull/68), który dokładnie zgłębia nasze motywy oraz rzuca dodatkowe światło na konkretne decyzje projektowe i obecny stan techniki.
```

> **Co najważniejsze, Hooki działają równolegle z istniejącym kodem, więc możesz wdrażać je stopniowo.** Nie ma pośpiechu, aby migrować kod do Hooków. Zalecamy unikanie "wielkiego przepisywania", szczególnie dla istniejących, skomplikowanych komponentów klasowych. Potrzeba delikatnie przestawić myślenie, aby zacząć "myśleć w Hookach". Z naszego doświadczenia wynika, że najlepiej poćwiczyć używanie Hooków na nowych, niekrytycznych komponentach i upewnić się, że wszyscy członkowie zespołu czują się z nimi komfortowo. Po wypróbowaniu Hooków, prosimy; [prześlij nam opinię](https://github.com/facebook/react/issues/new). Zarówno pozytywną, jak i negatywną.

Reviewer: ```suggestion
**Co najważniejsze, hooki działają równolegle z istniejącym kodem, więc możesz wdrażać je stopniowo.** Nie ma pośpiechu, aby migrować kod do Hooków. Zalecamy unikanie "wielkiego przepisywania", szczególnie dla istniejących, skomplikowanych komponentów klasowych. Potrzeba delikatnie przestawić myślenie, aby zacząć "myśleć hookami". Z naszego doświadczenia wynika, że najlepiej poćwiczyć używanie hooków na nowych, niekrytycznych komponentach i upewnić się, że wszyscy członkowie zespołu czują się z nimi komfortowo. Po wypróbowaniu hooków, prosimy - [prześlij nam opinię](https://github.com/facebook/react/issues/new). Zarówno pozytywną, jak i negatywną.
```

> Chcielibyśmy żeby Hooki objęły wszystkie możliwe przypadki użycia klas, ale **w możliwej do przewidzenia przyszłości, będziemy kontynuować wsparcie komponentów klasowych.** W Facebooku mamy dziesiątki tysięcy komponentów napisanych, jako klasy i stanowczo nie planujemy ich przepisywania. Zamiast tego, zaczynamy używać Hooków w nowym kodzie, równolegle do klas.

Reviewer: ```suggestion
Chcielibyśmy, żeby Hooki objęły wszystkie możliwe przypadki użycia klas, ale **w możliwej do przewidzenia przyszłości, będziemy kontynuować wsparcie komponentów klasowych.** W Facebooku mamy dziesiątki tysięcy komponentów napisanych jako klasy i zdecydowanie nie planujemy ich przepisywania. Zamiast tego zaczynamy używać Hooków w nowym kodzie, równolegle do klas.
```

## Najczęściej zadawane pytania {#frequently-asked-questions}

> Przygotowaliśmy rozdział [Hooki - FAQ](/docs/hooks-faq.html), który odpowiada na najczęściej zadawane pytania.

Reviewer: ```suggestion
Przygotowaliśmy rozdział pt. ["Hooki - FAQ"](/docs/hooks-faq.html), który odpowiada na najczęściej zadawane pytania.
```

## Kolejne kroki {#next-steps}

> Po przeczytaniu tego rozdziału powinieneś już wiedzieć, jakie problemy rozwiązują Hooki, ale wiele szczegółów jest prawdopodobnie niejasnych. Nie martw się! **Przejdźmy do [następnego rozdziału](/docs/hooks-overview.html), gdzie zaczniemy naukę o Hookach na przykładach.**

Reviewer: ```suggestion
Po przeczytaniu tego rozdziału powinien ukształtować ci się obraz tego, jakie problemy rozwiązują hooki, ale wiele szczegółów jest jeszcze prawdopodobnie niejasnych. Nie martw się! **Przejdźmy do [następnego rozdziału](/docs/hooks-overview.html), gdzie zaczniemy naukę o hookach na przykładach.**
```
